templates/template.java

```diff
@@ -0,0 +1,139 @@
+import java.util.stream.Stream;
+
+/**
+ * {filename} for Unicode {unicode_version}
+ * See <a href="https://github.com/ridiculousfish/widecharwidth/">https://github.com/ridiculousfish/widecharwidth/</a>
+ *
+ * <p>SHA1 file hashes:
+ * (
+ * the hashes for generate.py and the template are git object hashes,
+ * use `git log --all --find-object=hash` in the widecharwidth repository
+ * to see which commit they correspond to,
+ * or run `git hash-object` on the file to compare.
+ * The other hashes are simple `sha1sum` style hashes.
+ * )
+ *
+ * <ul>
+ * <li>generate.py:         {generate_hash}</li>
+ * <li>template.java:       {template_hash}</li>
+ * <li>UnicodeData.txt:     {unicode_hash}</li>
+ * <li>EastAsianWidth.txt:  {eaw_hash}</li>
+ * <li>emoji-data.txt:      {emoji_hash}</li>
+ * </ul>
+ */
+public class WcWidth {{
+
+    private WcWidth() {{
+    }}
+
+    public enum Type {{
+        ONE(1, ASCII_TABLE),                                    // The character is single-width
+        PRIVATE_USE(1, PRIVATE_TABLE),                          // The character is for private use.
+        NON_PRINT(0, NON_PRINT_TABLE),                          // The character is not printable
+        NON_CHARACTER(0, NON_CHARACTER_TABLE),                  // The character is a non-character.
+        COMBINING(0, COMBINING_TABLE, COMBINING_LETTERS_TABLE), // The character is a zero-width combiner
+        TWO(2, DOUBLE_WIDE_TABLE),                              // The character is double-width
+        AMBIGUOUS(1, AMBIGUOUS_TABLE),                          // The character is East-Asian ambiguous width.
+        UNASSIGNED(0, UNASSIGNED_TABLE),                        // The character is unassigned.
+        WIDENED_IN_9(2, WIDENED_IN_9_TABLE),                    // Width is 1 in Unicode 8, 2 in Unicode 9+.
+        UNKNOWN(1);
```

Comment:
I can throw an exception, but I'd prefer to return some value and let the users decide what to do in that case since it will make adding fallbacks if no errors are to be thrown easier. How about naming it `NOT_A_CODE_POINT`?

Reply:
This is a programming error. If you pass a value outside of this range, you *definitely* did something wrong before. It's not that this is a "weird character" that you can ignore or handle, it's that this a value that can't be encoded. That means there is no way to get it by decoding a string, and there is nothing you can do to turn it into a string.

It's a bug, and it should be an actual error, and in java I believe that's an exception. This also matches what we do in the python version. The others seem to just ignore the possibility, which is also an option in a "garbage in, garbage out" kind of sense. (also in C++ we don't have a good way of signalling an error since we don't want to enable exceptions)